File: sdk/go/integration.go
package merchantplatform

import (
	"context"
	"errors"
	"math/rand"
	"regexp"
	"strconv"
	"time"
)

type Environment string

const (
	Live    Environment = "live"
	Sandbox Environment = "sandbox"
)

type EndpointConfig struct {
	Environment Environment
	BaseURL     string
}

func LiveEndpoint(baseURL string) EndpointConfig    { return EndpointConfig{Live, baseURL} }
func SandboxEndpoint(baseURL string) EndpointConfig { return EndpointConfig{Sandbox, baseURL} }

type TelemetryEvent struct {
	Phase, Operation, Method string
	Status                   int
	Duration                 time.Duration
	Retryable                bool
}
type TelemetryHook func(TelemetryEvent)

func Instrument[T any](operation, method string, hook TelemetryHook, action func() (T, error)) (T, error) {
	if !regexp.MustCompile(`^[a-z][a-z0-9_.-]{0,63}$`).MatchString(operation) || !regexp.MustCompile(`^[A-Z]{3,7}$`).MatchString(method) {
		var zero T
		return zero, errors.New("telemetry operation or method is not low-cardinality")
	}
	started := time.Now()
	if hook != nil {
		hook(TelemetryEvent{Phase: "start", Operation: operation, Method: method})
	}
	value, err := action()
	event := TelemetryEvent{Phase: "end", Operation: operation, Method: method, Status: 200, Duration: time.Since(started)}
	var api *APIError
	if errors.As(err, &api) {
		event.Status = api.Status
		event.Retryable = api.Retryable
	}
	if hook != nil {
		hook(event)
	}
	return value, err
}

type RetryPolicy struct {
	MaxAttempts         int
	BaseDelay, MaxDelay time.Duration
	JitterRatio         float64
}

func DefaultRetryPolicy() RetryPolicy {
	return RetryPolicy{4, 200 * time.Millisecond, 5 * time.Second, 0.2}
}
func WithRetry[T any](ctx context.Context, safe bool, idempotencyKey string, policy RetryPolicy, action func() (T, error)) (T, error) {
	var zero T
	if !safe && idempotencyKey == "" {
		return zero, errors.New("unsafe retries require an idempotency key")
	}
	if policy.MaxAttempts < 1 || policy.MaxAttempts > 10 {
		return zero, errors.New("max attempts must be 1..10")
	}
	for attempt := 1; ; attempt++ {
		value, err := action()
		if err == nil {
			return value, nil
		}
		var api *APIError
		if attempt >= policy.MaxAttempts || !errors.As(err, &api) || !api.Retryable {
			return zero, err
		}
		delay := policy.BaseDelay * time.Duration(1<<(attempt-1))
		if delay > policy.MaxDelay {
			delay = policy.MaxDelay
		}
		if api.RetryAfter > 0 {
			delay = api.RetryAfter
		} else {
			factor := 1 + (rand.Float64()*2-1)*policy.JitterRatio
			delay = time.Duration(float64(delay) * factor)
		}
		if delay > policy.MaxDelay {
			delay = policy.MaxDelay
		}
		timer := time.NewTimer(delay)
		select {
		case <-ctx.Done():
			timer.Stop()
			return zero, ctx.Err()
		case <-timer.C:
		}
	}
}

func IteratePaymentIntents(ctx context.Context, client *Client, status string, pageSize int, yield func(PaymentIntent) error) error {
	after := ""
	for {
		page, err := client.ListPaymentIntents(ctx, status, after, pageSize)
		if err != nil {
			return err
		}
		for _, item := range page.Data.Items {
			if err = yield(item); err != nil {
				return err
			}
		}
		if page.Data.NextCursor == "" || page.Data.NextCursor == after {
			return nil
		}
		after = page.Data.NextCursor
	}
}
func IterateEvents(ctx context.Context, client *Client, afterSequence int64, pageSize int, yield func(PublicEvent) error) (string, error) {
	cursor := afterSequence
	for {
		page, err := client.ListEvents(ctx, cursor, pageSize)
		if err != nil {
			return "", err
		}
		for _, item := range page.Data.Items {
			if err = yield(item); err != nil {
				return "", err
			}
		}
		if len(page.Data.Items) == 0 {
			return page.Data.NextSequence, nil
		}
		next, err := strconv.ParseInt(page.Data.NextSequence, 10, 64)
		if err != nil || next < 0 {
			return "", errors.New("invalid event sequence")
		}
		if next == cursor {
			return page.Data.NextSequence, nil
		}
		cursor = next
	}
}

File: sdk/typescript/src/integration.ts
import type { MerchantClient, MerchantApiError } from "./client.js";
import type { PaymentIntentStatus, PublicEvent } from "./models.js";

export type APIEnvironment = "live" | "sandbox";
export interface EndpointConfig { environment: APIEnvironment; baseUrl: string }
export function liveEndpoint(baseUrl: string): EndpointConfig { return { environment: "live", baseUrl }; }
export function sandboxEndpoint(baseUrl: string): EndpointConfig { return { environment: "sandbox", baseUrl }; }

export interface TelemetryEvent { phase: "start" | "end"; operation: string; method: string; status?: number; duration_ms?: number; retryable?: boolean }
export type TelemetryHook = (event: Readonly<TelemetryEvent>) => void;
export async function instrument<T>(operation: string, method: string, hook: TelemetryHook | undefined, action: () => Promise<T>): Promise<T> {
  if (!/^[a-z][a-z0-9_.-]{0,63}$/.test(operation) || !/^[A-Z]{3,7}$/.test(method)) throw new TypeError("telemetry operation or method is not low-cardinality");
  const started = performance.now(); hook?.({ phase: "start", operation, method });
  try { const value = await action(); hook?.({ phase: "end", operation, method, status: 200, duration_ms: Math.round(performance.now() - started) }); return value; }
  catch (error) { const api = error as Partial<MerchantApiError>; hook?.({ phase: "end", operation, method, status: typeof api.status === "number" ? api.status : 0, duration_ms: Math.round(performance.now() - started), retryable: api.retryable === true }); throw error; }
}

export interface RetryPolicy { maxAttempts: number; baseDelayMs: number; maxDelayMs: number; jitterRatio: number }
export const defaultRetryPolicy: RetryPolicy = { maxAttempts: 4, baseDelayMs: 200, maxDelayMs: 5_000, jitterRatio: 0.2 };
/** Explicit retry wrapper. Unsafe mutations require an idempotency key; errors must opt in with retryable=true. */
export async function withRetry<T>(action: () => Promise<T>, options: { safe: boolean; idempotencyKey?: string; policy?: Partial<RetryPolicy>; random?: () => number; sleep?: (milliseconds: number) => Promise<void> }): Promise<T> {
  if (!options.safe && !options.idempotencyKey) throw new TypeError("unsafe retries require an idempotency key");
  const policy = { ...defaultRetryPolicy, ...options.policy }; if (policy.maxAttempts < 1 || policy.maxAttempts > 10) throw new TypeError("maxAttempts must be 1..10");
  const random = options.random ?? Math.random; const sleep = options.sleep ?? ((ms) => new Promise((resolve) => setTimeout(resolve, ms)));
  for (let attempt = 1;; attempt += 1) { try { return await action(); } catch (error) { const api = error as Partial<MerchantApiError>; if (attempt >= policy.maxAttempts || api.retryable !== true) throw error; const exponential = Math.min(policy.maxDelayMs, policy.baseDelayMs * 2 ** (attempt - 1)); const jittered = exponential * (1 + (random() * 2 - 1) * policy.jitterRatio); await sleep(Math.max(0, Math.min(api.retryAfterMs ?? jittered, policy.maxDelayMs))); } }
}

export async function* iteratePaymentIntents(client: MerchantClient, options: { status?: PaymentIntentStatus; pageSize?: number } = {}) { let after: string | undefined; do { const page = await client.listPaymentIntents({ status: options.status, after, limit: options.pageSize ?? 100 }); for (const item of page.data.items) yield item; after = page.data.next_cursor || undefined; } while (after); }
export async function* iterateEvents(client: MerchantClient, afterSequence = "0", pageSize = 100): AsyncGenerator<PublicEvent, string> { let cursor = afterSequence; for (;;) { const page = await client.listEvents(cursor, pageSize); for (const item of page.data.items) yield item; if (page.data.next_sequence === cursor || page.data.items.length === 0) return page.data.next_sequence; cursor = page.data.next_sequence; } }
